File: EaterNet/Eater/Services/CustomerService.cs
using Eater.Dtos;
using Eater.Dtos.Form;
using Eater.Exceptions;
using Eater.Interfaces;
using Eater.Models;
using Mapster;
using System.Security.Claims;

namespace Eater.Services
{
    public class CustomerService : ICustomerService
    {
        private readonly IHttpContextAccessor _httpContextAccessor;
        private readonly ICustomerRepository _customerRepository;
        private readonly ITokenService _tokenService;
        private readonly IPasswordEncoder _passwordEncoder;

        public CustomerService(IHttpContextAccessor httpContextAccessor, ICustomerRepository customerRepository, ITokenService tokenService, IPasswordEncoder passwordEncoder)
        {
            _httpContextAccessor = httpContextAccessor;
            _customerRepository = customerRepository;
            _tokenService = tokenService;
            _passwordEncoder = passwordEncoder;
        }

        public async Task<CustomerDto> GetProfile()
        {
            var userId = _httpContextAccessor.HttpContext!.User!.FindFirstValue(Constants.ClaimTypes.Id);
            var customer = await _customerRepository.GetById(userId!);
            if (customer == null)
            {
                throw new NotFoundException("User not found");
            }
            return customer.Adapt<CustomerDto>();
        }

        public async Task<AuthResultDto> Login(LoginForm form)
        {
            var customer = await _customerRepository.GetById(form.UserId);
            if (customer == null)
            {
                throw new UnauthorizedException("Bad Credentials");
            }

            if (!_passwordEncoder.Matches(form.Password, customer.Password))
            {
                throw new UnauthorizedException("Bad Credentials");
            }

            return new AuthResultDto(
                accessToken: _tokenService.GenerateToken(customer),
                profile: customer.Adapt<CustomerDto>()
            );
        }

        public async Task<AuthResultDto> Register(RegisterForm form)
        {
            var existingCustomer = await _customerRepository.GetById(form.UserId);
            if (existingCustomer != null)
            {
                throw new ConflictException("User already exists");
            }

            var customer = new Customer(form.UserId, form.Name, _passwordEncoder.Encode(form.Password));
            _customerRepository.Save(customer);
            await _customerRepository.SaveChangesAsync();

            return new AuthResultDto(
                accessToken: _tokenService.GenerateToken(customer),
                profile: customer.Adapt<CustomerDto>()
            );
        }
    }
}

File: EaterSpring/src/main/kotlin/xyz/purema/eater/services/CustomerService.kt
package xyz.purema.eater.services

import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import xyz.purema.eater.dtos.AuthResultDto
import xyz.purema.eater.dtos.CustomerDto
import xyz.purema.eater.dtos.form.LoginForm
import xyz.purema.eater.dtos.form.RegisterForm
import xyz.purema.eater.exceptions.ConflictException
import xyz.purema.eater.exceptions.NotFoundException
import xyz.purema.eater.exceptions.UnauthorizedException
import xyz.purema.eater.interfaces.ICustomerService
import xyz.purema.eater.interfaces.IMapper
import xyz.purema.eater.interfaces.ITokenService
import xyz.purema.eater.models.Customer
import xyz.purema.eater.repositories.ICustomerRepository

@Service
class CustomerService(
    private val mapper: IMapper,
    private val passwordEncoder: PasswordEncoder,
    private val customerRepository: ICustomerRepository,
    private val tokenService: ITokenService
) : ICustomerService {
    override fun register(form: RegisterForm): AuthResultDto {
        customerRepository.findById(form.userId!!).ifPresent {
            throw ConflictException("User already exists")
        }

        val customer = Customer(form.userId!!, form.name!!, passwordEncoder.encode(form.password!!))
        customerRepository.save(customer)

        return AuthResultDto(
            accessToken = tokenService.generateToken(customer),
            profile = mapper.mapCustomer(customer)
        )
    }

    override fun login(form: LoginForm): AuthResultDto {
        val customer = customerRepository.findById(form.userId!!).orElseThrow {
            UnauthorizedException("Bad Credentials")
        }

        if (!passwordEncoder.matches(form.password!!, customer.password)) {
            throw UnauthorizedException("Bad Credentials")
        }

        return AuthResultDto(
            accessToken = tokenService.generateToken(customer),
            profile = mapper.mapCustomer(customer)
        )
    }

    override fun getProfile(): CustomerDto {
        val userId = SecurityContextHolder.getContext().authentication.principal as String
        val customer = customerRepository.findById(userId).orElseThrow {
            NotFoundException("User not found")
        }

        return mapper.mapCustomer(customer)
    }
}